File: 0035-search-insert-position/0035-search-insert-position.py
class Solution(object):
    def searchInsert(self, nums, target):
        def binarySearch(arr, left, right, a):
            if right >= left:
                mid = left + (right - left)//2
                if arr[mid] == a:
                    return mid
                elif arr[mid] > a:
                    return binarySearch(arr, left, mid - 1, a)
                else:
                    return binarySearch(arr, mid + 1, right, a)
            else:
                return left

        return binarySearch(nums, 0, len(nums) - 1, target)

File: 0035-search-insert-position/0035-search-insert-position.cpp
class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        int l = 0, r = nums.size() - 1;

        function<int(int, int)> uppBound = [&](int left, int right) {
            if (left <= right) {
                int mid = left + (right - left) / 2;

                if (nums[mid] >= target) {
                    return uppBound(left, mid - 1);
                } else {
                    return uppBound(mid + 1, right);
                }
            } else {
                return right + 1;
            }
        };

        return uppBound(l, r);
    }
};